dlt/extract/concurrency.py

```diff
@@ -0,0 +1,219 @@
+        self._async_pool: asyncio.AbstractEventLoop = None
+        self._async_pool_thread: Thread = None
+        self.workers = workers
+        self.poll_interval = poll_interval
+        self.max_parallel_items = max_parallel_items
+        self.used_slots: int = 0
+
+    def __len__(self) -> int:
+        return len(self.futures)
+
+    @property
+    def free_slots(self) -> int:
+        # Done futures don't count as slots, so we can still add futures
+        return self.max_parallel_items - self.used_slots
+
+    @property
+    def empty(self) -> bool:
+        return len(self.futures) == 0
+
+    def _ensure_thread_pool(self) -> ThreadPoolExecutor:
+        # lazily start or return thread pool
+        if self._thread_pool:
+            return self._thread_pool
+
+        self._thread_pool = ThreadPoolExecutor(
+            self.workers, thread_name_prefix=Container.thread_pool_prefix() + "threads"
+        )
+        return self._thread_pool
+
+    def _ensure_async_pool(self) -> asyncio.AbstractEventLoop:
+        # lazily create async pool is separate thread
+        if self._async_pool:
+            return self._async_pool
+
+        def start_background_loop(loop: asyncio.AbstractEventLoop) -> None:
+            asyncio.set_event_loop(loop)
+            loop.run_forever()
+
+        self._async_pool = asyncio.new_event_loop()
+        self._async_pool_thread = Thread(
+            target=start_background_loop,
+            args=(self._async_pool,),
+            daemon=True,
+            name=Container.thread_pool_prefix() + "futures",
+        )
+        self._async_pool_thread.start()
+
+        # start or return async pool
+        return self._async_pool
+
+    def _vacate_slot(self, _: TItemFuture) -> None:
+        # Used as callback to free up slot when future is done
+        self.used_slots -= 1
+
+    def submit(self, pipe_item: ResolvablePipeItem) -> Optional[TItemFuture]:
+        """Submit an item to the pool.
+
+        Args:
+            pipe_item: The item to submit.
+
+        Returns:
+            The future if the item was successfully submitted, otherwise None.
+        """
+        if self.free_slots == 0:
+            return None
+
+        if self.free_slots < 0:  # TODO: Sanity check during dev, should never happen
+            raise RuntimeError("Used more slots than max_parallel_items")
+
+        # submit to thread pool or async pool
+        item = pipe_item.item
+        if isinstance(item, Awaitable):
+            future = asyncio.run_coroutine_threadsafe(item, self._ensure_async_pool())
+        elif callable(item):
+            future = self._ensure_thread_pool().submit(item)
+        else:
+            raise ValueError(f"Unsupported item type: {type(item)}")
+
+        # Future is not removed from self.futures until it's been consumed by the
+        # pipe iterator. But we always want to vacate a slot so new jobs can be submitted
+        future.add_done_callback(self._vacate_slot)
```

Comment:
why you call it then? it just decreases the counter

Reply:
It's the same idea (I think) as the way it checked for free slots before:
```python
                if len(self._futures) < self.max_parallel_items or self._next_future() >= 0:
```
But the difference now there can be more than `max_parallel_items + 1` futures in the pool if some of them are done, so this keeps track of how many slots are being blocked by pending futures only.  
The callback fires soon as the future is done which can be way before pipe iterator actually consumes the result and this seems faster than counting slots every time.